source/fuzz/fuzzer_pass_replace_params_with_struct.cpp

```diff
@@ -0,0 +1,106 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/fuzzer_pass_replace_params_with_struct.h"
+
+#include <numeric>
+#include <vector>
+
+#include "source/fuzz/fuzzer_context.h"
+#include "source/fuzz/fuzzer_util.h"
+#include "source/fuzz/transformation_replace_params_with_struct.h"
+
+namespace spvtools {
+namespace fuzz {
+
+FuzzerPassReplaceParamsWithStruct::FuzzerPassReplaceParamsWithStruct(
+    opt::IRContext* ir_context, TransformationContext* transformation_context,
+    FuzzerContext* fuzzer_context,
+    protobufs::TransformationSequence* transformations)
+    : FuzzerPass(ir_context, transformation_context, fuzzer_context,
+                 transformations) {}
+
+FuzzerPassReplaceParamsWithStruct::~FuzzerPassReplaceParamsWithStruct() =
+    default;
+
+void FuzzerPassReplaceParamsWithStruct::Apply() {
+  for (const auto& function : *GetIRContext()->module()) {
+    auto params =
+        fuzzerutil::GetParameters(GetIRContext(), function.result_id());
+
+    if (params.empty() || fuzzerutil::FunctionIsEntryPoint(
+                              GetIRContext(), function.result_id())) {
+      continue;
+    }
+
+    if (!GetFuzzerContext()->ChoosePercentage(
+            GetFuzzerContext()->GetChanceOfReplacingParametersWithStruct())) {
+      continue;
+    }
+
+    std::vector<uint32_t> parameter_id(params.size());
+    std::iota(parameter_id.begin(), parameter_id.end(), 0);
+
+    // Remove unsupported parameters.
+    auto new_end = std::remove_if(
+        parameter_id.begin(), parameter_id.end(),
+        [this, &params](uint32_t index) {
+          const auto* type =
+              GetIRContext()->get_type_mgr()->GetType(params[index]->type_id());
+          return !type || !TransformationReplaceParamsWithStruct::
+                              IsParameterTypeSupported(*type);
+        });
+
+    parameter_id.erase(new_end, parameter_id.end());
+
+    if (parameter_id.empty()) {
+      continue;
+    }
+
+    auto num_replaced_params = std::min<size_t>(
+        parameter_id.size(),
+        GetFuzzerContext()->GetRandomNumberOfParametersReplacedWithStruct(
+            static_cast<uint32_t>(params.size())));
+
+    GetFuzzerContext()->Shuffle(&parameter_id);
+    parameter_id.resize(num_replaced_params);
+
+    // Make sure OpTypeStruct exists in the module.
+    auto component_type_ids = parameter_id;
+    for (auto& id : component_type_ids) {
+      id = params[id]->type_id();
+    }
```

Comment:
I find this hard to read.  I would be much more comfortable with:

- making an empty vector, `component_type_ids`
- iterating through `parameter_id` (but with the name indicating that it represents indices)
- adding the relevant thing to `component_type_ids` on each iteration

In general I'm not keen on the:

for (auto& element_reference : collection) {
  // update element_reference
}

as a way to update a collection, because when you just look at the line of code that updates the element reference you can't see immediately what collection is being updated.

Also I don't like patterns that involve adding the *wrong* thing to a collection and then making it right; I prefer to start with nothing and then add the right things.  Again, this comes down to clarity for readers, and reducing the possibility for others who work on the code later to introduce errors.

Reply:
Indeed, you are right! Fixed in 97a33fa.